### ISSUE TITLE
 [Feature request] Scatter sort

### ISSUE THREAD

**Issue by soumyasanyal** [rusty1s/pytorch_scatter#48]
Hi!
Similar to scatter_max, can we have a scatter_sort that sorts tensors based on `index`? Here, `index` represents the graph to which a node belongs and `src` represents all the nodes in a batch consisting of multiple graphs.

**Comment by rusty1s** [comment]
Hi, this is not really a `scatter` op, but rather a `sparse_sort`, because the number of element stays the same. Hence, it does not really fit into the scope of this package. For implementing `topk` pooling in PyG, we implemented a `sparse_sort` op by creating dense matrices first, sorting them, and masking invalid entries out. This is not ideal, but fast enough for most use-cases, because the `sparse_sort` op is rather difficult to implement.

**Comment by tformal** [comment]
Hi @rusty1s ,
If I understand correctly:
* there is no easy way to extend `scatter_max` to `scatter_topk` ?
* a workaround is to use the `topk` function you implemented in `torch_geometric.nn.pool.topk_pool` ?

**Comment by rusty1s** [comment]
Yes and yes. A good alternative is the usage if the keops library, which comes with a memory efficient `argKmin` implementation.

**Comment by flandolfi** [comment]
Hi @rusty1s,

I was looking for something similar to what you refer as `sparse_sort`. Do you think this might be a valid approach?

```python
import torch

def sparse_sort(src: torch.Tensor, index: torch.Tensor, dim=0, descending=False, eps=1e-12):
    f_src = src.float()
    f_min, f_max = f_src.min(dim)[0], f_src.max(dim)[0]
    norm = (f_src - f_min)/(f_max - f_min + eps) + index.float()*(-1)**int(descending)
    perm = norm.argsort(dim=dim, descending=descending)

    return src[perm], perm
```

In practice:

1. we normalize the `src` in the range [0, 1);
2. we sum (or subtract, if we sort in descending order) the index value. In this way, the values of `src` for each index *i* will fall in the range [*i*, *i* + 1) (or, alternatively, [-*i*, -*i* + 1));
3. we `argsort` the results and obtain the final permutation.

The only drawback could be the numerical stability with wide ranges, but I think this could work in most cases.

I believe we could also extend it to a `sparse_topk`, creating a mask with something like
```python
count = index.bincount()
mask = torch.arange(0, index.size(0)) - count[index] + count[0]
mask = mask < k
```

Thank you so much for your library, anyway. I hope you and your relatives are all ok and safe.

Francesco

**Comment by rusty1s** [comment]
That looks like a totally valid solution!

**Comment by flandolfi** [comment]
A lexical sort (or two stable sorts) would be better, anyway! This is just an alternative until PyTorch will provide one of the two

**Comment by simonrouse9461** [comment]
@flandolfi Thanks for the brilliant solution! Here's an improved one that works with multidimensional tensors:
```python
def sparse_sort(src: torch.Tensor, index: torch.Tensor, dim=0, descending=False, eps=1e-12):
    f_src = src.float()
    f_min, f_max = f_src.min(dim=dim, keepdim=True).values, f_src.max(dim=dim, keepdim=True).values
    norm = (f_src - f_min) / (f_max - f_min + eps) + index.float() * (-1) ** int(descending)
    perm = list(torch.meshgrid(*[torch.arange(i).to(src.device) for i in src.shape], indexing='ij'))
    perm[dim] = norm.argsort(dim=dim, descending=descending)

    return src[perm], perm[dim]
```

**Comment by flandolfi** [comment]
Hi @simonrouse9461! Thanks for the update :) 

Honestly I think that this approach can be abandoned since newer versions of pytorch allow stable sorts, and hence it should be easier to compute sparse sorts.

Here you can find an example where sparse sort is obtained by a sequence of two stable sorts:
https://github.com/pyg-team/pytorch_geometric/blob/ecf40202a4f5aaeb264b7183cc5038fdf14a45ad/torch_geometric/nn/aggr/quantile.py#L88

**Comment by simonrouse9461** [comment]
@flandolfi Looks like a much simpler and more elegant solution. Thanks!
